android/space/src/main/kotlin/net/twisterrob/inventory/android/space/ManageSpaceUiStateHandler.kt

```diff
@@ -0,0 +1,43 @@
+package net.twisterrob.inventory.android.space
+
+import net.twisterrob.android.utils.tools.DialogTools
+import net.twisterrob.inventory.android.arch.UiStateHandler
+import net.twisterrob.inventory.android.space.ManageSpaceUiState.ConfirmationUiState
+import net.twisterrob.inventory.android.space.databinding.ManageSpaceActivityBinding
+import javax.inject.Inject
+
+internal class ManageSpaceUiStateHandler @Inject constructor(
+	private val binding: ManageSpaceActivityBinding,
+	private val viewModel: ManageSpaceViewModel,
+) : UiStateHandler<ManageSpaceUiState> {
+	override fun render(state: ManageSpaceUiState) {
+		binding.refresher.isRefreshing = state.isLoading
+		// Note: isVisible is set from XML based on debug/release builds for each button/section.
+		binding.contents.storageImageCacheActions.isEnabled = state.isEnableButtons
+		binding.contents.storageDbActions.isEnabled = state.isEnableButtons
+		binding.contents.storageDataActions.isEnabled = state.isEnableButtons
+		binding.contents.storageIndexActions.isEnabled = state.isEnableButtons
+		binding.contents.storageImageCacheSize.text = state.sizes?.imageCache
+		binding.contents.storageDbSize.text = state.sizes?.database
+		binding.contents.storageDbFreelistSize.text = state.sizes?.freelist
+		binding.contents.storageSearchSize.text = state.sizes?.searchIndex
+		binding.contents.storageAllSize.text = state.sizes?.allData
+		if (state.confirmation != null) {
+			showDialog(state.confirmation)
+		}
+	}
+
+	private fun showDialog(confirmation: ConfirmationUiState) {
```

Comment:
What is the thinking behind modelling the dialog in the state rather than as an effect?

Reply:
Short answer: Config changes.

It's a modal confirmation dialog (as opposed to an informational forgettable "OK" dialog), I want the user to answer the question before continuing, rotating the device shouldn't reset the user flow.

If you think about the dialog as being just a `FrameLayout` overlay, then you would almost _have to_ model it this way. Having a separate API (compared to Views) was done way before any architecture came about for Android. This solution is really quite detached from the implementation details of the Android Dialog API, it could be showing a SnackBar, or an inline dismissible item too. The information where to go next is also not stored in the dialog's listeners, but instead in the ViewModel (which outlives the View). See `ManageSpaceViewModel.confirmedClean`'s `when (confirmations.receive()) {` line. This means the View can re-attach to the VM mid-confirmation and continue code after receiving the button press from UI. Note that this needs a tiny bit of forethought in `onResume` (`screenVisible`'s `if (state.isModalShowing) {`).

Thoughts?

Reply:
Nothing insightful apart from sounds sensible to me! Thanks for the explanation